modules/stochastic_tools/src/surrogates/LibtorchDRLControlTrainer.C

```diff
@@ -0,0 +1,498 @@
+    convertDataToTensor(_log_probability_data, _log_probability_tensor);
+    // Discard (detach) the gradient info for return data
+    convertDataToTensor(_return_data, _return_tensor, true);
+
+    // We train the controller using the emulator to get a good control strategy
+    trainController();
+
+    // We clean the training data after contoller update and reset the counter
+    resetData();
+  }
+#endif
+}
+
+void
+LibtorchDRLControlTrainer::computeAverageEpisodeReward()
+{
+  if (_reward_data.size())
+    _average_episode_reward =
+        std::accumulate(_reward_data.begin(), _reward_data.end(), 0.0) / _reward_data.size();
+}
+
+void
+LibtorchDRLControlTrainer::getInputDataFromReporter(
+    std::vector<std::vector<Real>> & data,
+    const std::vector<ReporterName> & reporter_names,
+    unsigned int num_timesteps)
+{
+  for (unsigned int rep_i = 0; rep_i < reporter_names.size(); rep_i++)
+  {
+    std::vector<Real> reporter_data =
+        getReporterValueByName<std::vector<Real>>(reporter_names[rep_i]);
+
+    // We shift and scale the inputs to get better training efficiency
+    std::transform(
+        reporter_data.begin(),
+        reporter_data.end(),
+        reporter_data.begin(),
+        [this, &rep_i](Real value) -> Real
+        { return (value - _response_shift_factors[rep_i]) * _response_scaling_factors[rep_i]; });
+
+    // Fill the corresponding containers
+    for (unsigned int start_step = 0; start_step < num_timesteps; start_step++)
+    {
+      unsigned int row = reporter_names.size() * start_step + rep_i;
+      for (unsigned int fill_i = 1; fill_i < num_timesteps - start_step; ++fill_i)
+        data[row].push_back(reporter_data[0]);
+
+      data[row].insert(data[row].end(),
+                       reporter_data.begin(),
+                       reporter_data.begin() + start_step + reporter_data.size() -
+                           (num_timesteps - 1) - _shift_outputs);
+    }
+  }
+}
+
+void
+LibtorchDRLControlTrainer::getOutputDataFromReporter(
+    std::vector<std::vector<Real>> & data, const std::vector<ReporterName> & reporter_names)
+{
+  for (unsigned int rep_i = 0; rep_i < reporter_names.size(); rep_i++)
+  {
+    const std::vector<Real> & reporter_data =
+        getReporterValueByName<std::vector<Real>>(reporter_names[rep_i]);
+
+    // Fill the corresponding containers
+    data[rep_i].insert(
+        data[rep_i].end(), reporter_data.begin() + _shift_outputs, reporter_data.end());
+  }
+}
+
+void
+LibtorchDRLControlTrainer::getRewardDataFromReporter(std::vector<Real> & data,
+                                                     const ReporterName & reporter_name)
+{
+  const std::vector<Real> & reporter_data =
+      getReporterValueByName<std::vector<Real>>(reporter_name);
+
+  // Fill the corresponding container
+  data.insert(data.end(), reporter_data.begin() + _shift_outputs, reporter_data.end());
+}
+
+#ifdef LIBTORCH_ENABLED
+void
+LibtorchDRLControlTrainer::computeDiscountedRewards()
```

Comment:
`Reward-To-Go` is the acknowledged name in DRL community. The reward-to-go computed here is not only `discounted` by the `decaying_factor` at every step, but also a summation over all of them. So I'd prefer to switch back the name here

Reply:
Will change it back with the next commit, but I don't think reward-to-go is indicative of what is happening under the hood.

Reply:
I agree it is a bad name but people keep using it for some reason